Make Schema conform to Table Schema Spec.

vlandham opened the pull request:
This changes the `Schema` interface to be compliant with the Table Schema spec. 

https://specs.frictionlessdata.io/table-schema/

As discussed in https://github.com/vega/voyager/issues/338, here we:

* provide interfaces for `TableSchemaFieldDescriptor` and `TableSchema`
* rename field to name - the required ID of the field
* rename type to vlType - as this is additional data not captured in table schema
* rename primitiveType to type
* rename primitiveType date to datetime.

This introduces some inconsistencies between this new `Schema` and (for example) the `FieldQuery` interface (field === name). we can work toward changing more interfaces if this is a problem.

kanitw commented:
Wow this is fast -- looks like the test is failing though.   

I'll take a look after I arrive at UW (still commuting)

vlandham commented:
haha. yeah, i was working on it - hoping it would be agreed upon ;) - and yes - looks like i was mistaken about where compassql was running tests. will fix that now.

vlandham commented on src/schema.ts:
```diff
@@ -131,44 +164,44 @@ export function build(data: any, opt: QueryConfig = {}): Schema {
 
   /** @return a list of FieldSchemas */
-  public get fieldSchemas() {
+  public get fields() {
```
The `TableSchema` spec requires an array called `fields` - so thats why i changed it:

https://specs.frictionlessdata.io/table-schema/#descriptor

kanitw replied:
Now I wonder why we can override `fields` array as a getter.  This seems to cloak the `fields` that comes with TableSchema.  

Given this and [the comment here](https://github.com/vega/compassql/pull/379/files#r121759269), I start wondering if we better make Schema encapsulate TableSchema rather than `implements` it as discussed [here](https://github.com/vega/compassql/pull/379/files#r121749686)?

kanitw replied:
Ok now I understand how it works better you can ignore this comment if we chose to do `implements` rather than composition.  (If we do composition, then I still prefer the original name.)

kanitw commented on src/schema.ts:
```diff
@@ -131,44 +164,44 @@ export function build(data: any, opt: QueryConfig = {}): Schema {
 }
 
-export class Schema {
+export class Schema implements TableSchema {
```
Also, when I try

```
interface AI {
  a: string;
}

class A implements AI {
  
}
```
I got an error that `A` doesn't have `a: string`. 

I'm not sure why there is no similar error for schema here.

vlandham replied:
I believe its because `Schema` does indeed implement the required field of `TableSchema` - which is `fields`

kanitw replied:
Ah! Just noticed that other properties are optional.

kanitw replied:
But this means that we cannot access other properties of TableSchema via Schema anyway?

Here is a small code that I experiment with:
```
interface AI {
  a: string;
  b?: number;
}

class A implements AI {
  public a: string;
}

const aa = new A();
aa.a = 'a';
aa.b = 1; // error Property b does not exist on type A
```

vlandham replied:
Right - cause we are not extending it - but indicating that Schema conforms to the interface. If there is optional values in the interface - we wouldn't expect to always have those optional values.

kanitw replied:
Yeah, but it's kinda weird that we don't make them accessible.  

Now I start liking encapsulation more again.

If `tableSchema: ReadOnly<TableSchema<FieldSchema>>` is a property of Schema, then we can easily export the TableSchema object.  In addition, we can directly access all of the TableSchema via Schema?

Note that Schema's constructor is only called via the `build()` method, so we have the freedom to modify it however we like.

vlandham replied:
in this scenario - what is the argument of TableSchema being parameterizable with `<FieldSchema>` ? In order to _be_ Table Schema, it needs to hold fields of a particular type. I believe the idea that `FieldSchema` **extends** `TableSchemaFieldDescriptor` would still work well in this encapsulated approach - and would simplify the signature of these interfaces.

kanitw replied:
> in this scenario - what is the argument of TableSchema being parameterizable with <FieldSchema> ?

If we make it generic, we could do.  

```
interface TableSchema<F extends TableSchemaFieldDescriptor> {
   field: F;
   ...
}
```

kanitw commented:
> This introduces some inconsistencies between this new Schema and (for example) the FieldQuery interface (field === name). we can work toward changing more interfaces if this is a problem.

We definitely should keep `field` in `FieldQuery` as it should be consistent with Vega-Lite.

vlandham commented on src/schema.ts:
```diff
@@ -27,6 +46,17 @@ export interface FieldSchema {
+ * Table Schema
+ * see: https://specs.frictionlessdata.io/table-schema/
+ */
+export interface TableSchema {
```
i think we have figured this out below - but just to clarify - `_fieldSchemas` is not redundant - it is just implementing the interface in a particular way.

vlandham commented on src/constraint/field.ts:
```diff
@@ -217,13 +217,13 @@ export const FIELD_CONSTRAINTS: EncodingConstraintModel<FieldQuery>[] = [
       switch (primitiveType) {
-        case PrimitiveType.BOOLEAN:
-        case PrimitiveType.STRING:
+        case DataType.BOOLEAN:
```
okay

kanitw commented:
cc: @AkshatSh   Note that this PR may cause some conflicts with your #372

vlandham commented:
@kanitw I believe this new commit addresses the issues you brought up.

* Now Schema contains a dataTableSchema - and uses it as the main schema storage.
* the `build()` function can now take a dataTable schema - which will serve as the basis of the dataTableSchema used by `Schema`. 
* Tests have been added to define the interaction between the existing inferring of types and the input dataTable (input dataTable trumps inferred types).
* primitiveType is back

kanitw commented on src/schema.ts:
```diff
@@ -40,36 +71,44 @@ export interface FieldSchema {
  * @return a Schema object
  */
-export function build(data: any, opt: QueryConfig = {}): Schema {
+export function build(data: any, opt: QueryConfig = {}, tableSchema: TableSchema<TableSchemaFieldDescriptor> = {fields:[]}): Schema {
```
Maybe opt should remain the last parameter?

kanitw commented on src/schema.ts:
```diff
@@ -8,25 +8,56 @@ import * as dlBin from 'datalib/src/bins/bins';
 
-  title?: string;
   index?: number;
+  inIndex?: number;
```
Maybe `originalIndex` is more descriptive?

kanitw replied:
(`inIndex` confused me for a while)

kanitw commented on src/schema.ts:
```diff
@@ -128,47 +174,63 @@ export function build(data: any, opt: QueryConfig = {}): Schema {
   }
 
-  return new Schema(fieldSchemas);
+  const derivedTableSchema: TableSchema<FieldSchema> = {
```
```
const derivedTableSchema: TableSchema<FieldSchema> = {
   ...tableSchema,
   fields: fieldSchemas
};
```
would be shorter and also support future additional properties in the table schema?

kanitw commented on src/schema.ts:
```diff
@@ -78,17 +117,24 @@ export function build(data: any, opt: QueryConfig = {}): Schema {
-      field: field,
+    let fieldSchema = {
+      name: name,
+      inIndex: inIndex,
```
Add `// Need to keep original index for re-exporting TableSchema`

kanitw commented on src/schema.ts:
```diff
@@ -40,36 +71,44 @@ export interface FieldSchema {
+  let fieldSchemas: FieldSchema[] = summaries.map(function(fieldProfile, index) {
+
+    const name: string = fieldProfile.field;
+    const inIndex: number = index;
```
Since inIndex is only used once below in `let fieldSchema` -- just refer to `index` directly there?

vlandham commented:
@kanitw good suggestions. i believe i cover them all in the latest commit.

kanitw commented:
Thanks! :)
